ebpf/xdp_ew.c

```diff
@@ -69,10 +70,49 @@
 
 #define LINUX_VERSION_CODE 263682
 
-struct vlan_hdr {
-    __u16	h_vlan_TCI;
-    __u16	h_vlan_encapsulated_proto;
-};
+static INLINE int get_sport(void *trans_data, void *data_end, __u8 protocol)
+{
+    struct tcphdr *th;
+    struct udphdr *uh;
+
+    switch (protocol) {
+        case IPPROTO_TCP:
+            th = (struct tcphdr *)trans_data;
+            if ((void *)(th + 1) > data_end) {
+                return -1;
+            }
+            return th->source;
+        case IPPROTO_UDP:
+            uh = (struct udphdr *)trans_data;
+            if ((void *)(uh + 1) > data_end) {
+                return -1;
+            }
+            return uh->source;
+        default:
+            return 0;
+    }
+}
+
+static INLINE int get_dport(void *trans_data, void *data_end, __u8 protocol)
+{
+    struct tcphdr *th;
+    struct udphdr *uh;
+
+    switch (protocol) {
+        case IPPROTO_TCP:
+            th = (struct tcphdr *)trans_data;
+            if ((void *)(th + 1) > data_end)
+                return -1;
+            return th->dest;
+        case IPPROTO_UDP:
+            uh = (struct udphdr *)trans_data;
+            if ((void *)(uh + 1) > data_end)
+                return -1;
+            return uh->dest;
+        default:
+            return 0;
+    }
+}
 
 static int INLINE filter_ipv4(struct xdp_md *ctx, void *data, __u64 nh_off, void *data_end)
 {
```

Comment:
Should these be check for 0 to avoid null pointer dereferences? I don't know the caller here and whether is validated before reaching here

Reply:
This whole program gets compiled to BPF machine language which is byte-code scrutinized by the kernel for those sort of things, so I think we're okay.
The kernel validator is specifically cautious of (and will not a load a program that performs) memory dereferences that don't previously check whether they're valid.
In this case, `data` is the packet buffer (which the kernel has already validated to be non-zero) and data_end is the end of the packet (also already validated to be non-zero).
Each dereference of *(data + n), however, must be proceeded by a "if data + n > data_end then fail" guard, else the bpf validator wont load the program.